### ISSUE TITLE
WebDataset created in Python 3.8 failing to read

### ISSUE THREAD

**Issue by austinmw** [NVIDIA/DALI#3640]
Hi, I created a MNIST webdataset programmatically in Python following the example notebook here: https://github.com/webdataset/webdataset/blob/master/notebooks/creating.ipynb

I noticed that when I create shards of 1000 samples using Python webdataset, I get: a `4.4M POSIX tar archive`. This tar then fails to process via the DALI webdataset example here: https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/dataloading_webdataset.html

However, if I very simply just untar and retar the above, I then end up with a `2.4M POSIX tar archive (GNU)`, which can be processed by the DALI tutorial successfully.

So my question is, why does simply un-taring and re-taring change the file from 4.4M to 2.4M, add GNU label, and allow it to be read in successfully?

**edit:** I discovered the problem. It seems `fn.readers.webdataset` will only read GNU format tar files. However, in Python 3.8 `tarfile.DEFAULT_FORMAT` switched from `tarfile.GNU_FORMAT` to `tarfile.PAX_FORMAT`.  Would you please mind adding support for PAX format tar files so that people using Python 3.8 don't run into compatibility issues?

-----

Tar creation code for reference:

```python
def write_mnist_wd(base="./shards", download=True, maxsize=1e9, maxcount=100000):

    assert maxsize > 10000000
    assert maxcount < 1000000
    
    if not os.path.exists(base):
        os.makedirs(base)
        
    # def readfile(fname):
    #     "Read a binary file from disk."
    #     with open(fname, "rb") as stream:
    #         return stream.read()        
    
    # Process train then test
    for split, value in {'train': True, 'test': False}.items():
        ds = torchvision.datasets.MNIST(root="./temp", download=download, train=value)
        classes = ds.classes
        nimages = len(ds.data)
        indexes = list(range(nimages))
        #random.shuffle(indexes)        
        
        # This is the output pattern under which we write shards
        pattern = os.path.join(base, f"{split}-%06d.tar")
        
        with wds.ShardWriter(pattern, maxsize=int(maxsize), maxcount=int(maxcount)) as sink:
            
            all_keys = set()
            for i in tqdm(indexes, desc=f'Writing {split}'):
                
                #key = f'sample_{i:06d}'
                key = str(i)
                img, target = ds.data[i], int(ds.targets[i])              
                
                # Convert to PIL
                img = Image.fromarray(img.numpy(), mode='L')
                
                # Read the JPEG-compressed image file contents.
                #img = readfile('img.jpg')
                
                # Write compressed jpeg data to BytesIO object
                output = io.BytesIO()
                img.save(output, format='JPEG')
                
                # Get data
                img = output.getvalue()

                assert key not in all_keys
                all_keys.add(key)                
    
                # Construct a sample
                sample = {"__key__": key, "jpg": img, "cls": target}
            
                # Write the sample to the sharded tar archives
                sink.write(sample)
        
write_mnist_wd(maxcount=1000)
```

**Comment by JanuszL** [comment]
Hi @austinmw,

We will look into that and get back to you soon.

**Comment by austinmw** [comment]
Thanks! Also, much smaller question if you don't mind: 

Inside my `pipeline_def`, after I get `images_raw, labels = fn.readers.webdataset(...)`, how can I decode the `'cls'` extension int labels from ascii to their true values?

**Comment by JanuszL** [comment]
@austinmw,

You cannot do that inside `pipeline_def` as it provides a symbolic definition of the graph. You need to operate on the pipeline output directly.

**Comment by austinmw** [comment]
@JanuszL Hmmm alright. But it seems that I then have to copy from gpu to cpu, perform operation in numpy, then copy back to gpu. Is that right?

**Comment by JanuszL** [comment]
You don't have to store labels on the GPU side. You can keep it on the cpu, process in numpy and then copy to the GPU.

**Comment by JanuszL** [comment]
https://github.com/NVIDIA/DALI/pull/3645 should address the issue with PAX format.

**Comment by JanuszL** [comment]
@austinmw,
> Inside my pipeline_def, after I get images_raw, labels = fn.readers.webdataset(...), how can I decode the 'cls' extension int labels from ascii to their true values?

Another idea is to just encode the class as binary data, not ASCI value, then DALI can read it as it is without any conversion.

**Comment by JanuszL** [comment]
The fix has been merged. It will be available in the next nightly build and 1.11 release.

**Comment by JanuszL** [comment]
DALI 1.11 has been releases and the PAX issue should be addressed.
